test/e2e/common.go

```diff
@@ -35,15 +36,64 @@ type runOpts struct {
 	Redact       bool
 }
 
+var (
+	e env
+	k kn
+)
+
+const (
+	KnDefaultTestImage string = "gcr.io/knative-samples/helloworld-go"
+	MaxRetries         int    = 10
+)
+
+// Setup set up an enviroment for kn integration test returns the Teardown cleanup function
+func Setup(t *testing.T) func(t *testing.T) {
+	e = buildEnv(t)
+	e.Namespace = fmt.Sprintf("%s%d", e.Namespace, namespaceCount)
+	namespaceCount++
+	k = kn{t, e.Namespace, Logger{}}
+	CreateTestNamespace(t, e.Namespace)
+	return Teardown
+}
+
+// Teaddown clean up
+func Teardown(t *testing.T) {
+	DeleteTestNamespace(t, e.Namespace)
+}
+
 // CreateTestNamespace creates and tests a namesspace creation invoking kubectl
 func CreateTestNamespace(t *testing.T, namespace string) {
-	kubectl := kubectl{t, Logger{}}
-	out, err := kubectl.RunWithOpts([]string{"create", "namespace", namespace}, runOpts{})
-	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl create namespace' command. Error: %s", err.Error()))
+	logger := Logger{}
+	kubectlCreateNamespace := func() (string, error) {
+		kubectl := kubectl{t, logger}
+		return kubectl.RunWithOpts([]string{"create", "namespace", namespace}, runOpts{AllowError: true})
 	}
 
 	expectedOutputRegexp := fmt.Sprintf("namespace?.+%s.+created", namespace)
+	var (
+		created bool
+		retries int
+		out     string
+		err     error
+	)
+
+	for !created && retries < MaxRetries {
+		out, err = kubectlCreateNamespace()
+		if err != nil && retries >= MaxRetries {
```

Comment:
How can retries be larger than MaxRetries here (considering the loop condition) ?

Reply:
I think the logic of this loop is flawed.

I would put this in an extra function returning error or nil. So you could immediately return nil when success and an error after the loop, like in 

```golang
func createNamespace() error {

   retries := 0
   var err Error
   for retries < MaxRetries {
     _, err := kubectlCreateNamespace()
     if err == nil {
	    return nil
     } 
     time.Sleep(30 * time.Second)
     retries++
   }
   return err
}

```

Not tested, but hopefully good enough to clarify the point.

Reply:
It's more complicated than that but adding a helper private for the retry logic is fine. See update.